flow/project.py

```diff
@@ -479,24 +475,465 @@ def complete(self, job):
+        if hasattr(func, '_flow_group_operation_directives'):
+            if self.name in func._flow_group_operation_directives.keys():
+                raise ValueError("Attempted repeat directives setting of {} in group {}"
+                                 "".format(func, self.name))
+            else:
+                func._flow_group_operation_directives[self.name] = directives
+        else:
+            func._flow_group_operation_directives = {self.name: directives}
+
+    def with_directives(self, **kwargs):
+        """Returns a decorator that sets group specific directives to the operation.
+
+        :param kwargs:
+            Directives to use for resource requests and running the operation
+            through the group.
+        :type **kwargs:
+            dict
+        :returns:
+            A decorator which registers the function into the group with
+            specified directives.
+        :rtype:
+            function
+        """
+        directives = dict() if len(kwargs) == 0 else kwargs
+
+        def decorator(func):
+            self._set_directives(func, directives)
+            return self(func)
+
+        return decorator
+
+
+class FlowGroup(object):
+    """A FlowGroup represents a subset of a workflow for a project.
+
+    Any :py:class:`FlowGroup` is associated with a group of
+    :py:class:`FlowOperation`s. The command :py:param:`cmd` is a
+    :py:func:`FlowProject.run` command, but can be overwritten with any function
+    that takes job and operations arguments.
+
+    :param name:
+        The name of the group to be used when calling from the commandline.
+    :type name:
+        :class:`str`
+    :param operations:
+        A :class:`dict` of name : :class:`FlowOperation` key, value pairs in the
+        group.
+    :type operations:
+        :class:`dict` keys of :class:`str` and values of :class:`FlowOperation`
+    :param operation_directives:
+        A dictionary of additional parameters that provide instructions on how
+        to execute a particular operation, e.g., specifically required
+        resources.  Operation names are keys and the directive dictionary are
+        values. If an operation does not have directives specified, then the
+        directives of the singleton group containing that operation are used. To
+        prevent this set the directives to an empty dictionary for that
+        operation.
+    :type operation_directives:
+        :class:`dict`
+    :param options:
+        A string of options to append to the output of the object's call method.
+        This lets options like --num_passes to be given to a group.
+    :type options:
+        :class:`str`
+    """
+
+    MAX_LEN_ID = 100
+
+    def __init__(self, name, operations=None, operation_directives=None,
+                 options=None):
+        self.name = name
+        self.options = "" if options is None else options
+        self.operations = dict() if operations is None else operations
```

Comment:
In python, dict and list are mutable and should not be used for default parameters. This can lead to the default values being modified (e.g. key, value pairs being added to entrypoint). `None` is safer. However, strings can be used with no worries, so I will change that.

Reply:
Good catch, I forgot the consequences of doing that.